core/shared/src/main/scala/scalaz/zio/Semaphore.scala

```diff
@@ -0,0 +1,123 @@
+package scalaz.zio
+
+import scala.collection.immutable.{ Queue => IQueue }
+import scala.util.control.NoStackTrace
+
+abstract class Semaphore {
+
+  import Semaphore.NegativeArgument
+
+  //  def available[E]: IO[E, Long]
+
+  //  def count[E]: IO[E, Long]
+
+  def acquireN(n: Long): IO[NegativeArgument.type, Unit]
```

Comment:
If the user specifies a negative number, it's a bug in the code, I'd terminate the fiber. The other option is to treat < 0 as = 0, but that seems likely to mask bugs.